tests/func/test_ls.py

```diff
@@ -0,0 +1,153 @@
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+
+    files = Repo.ls(fspath(tmp_dir), recursive=True, outputs_only=True)
+    assert set(files) == set([os.path.join("dir1", "dvc_file2"), "dvc_file1"])
+
+
+def test_ls_repo_with_target(tmp_dir, dvc, scm):
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+
+    files = Repo.ls(fspath(tmp_dir), target="dir1", recursive=True)
+    assert set(files) == set(
+        [
+            os.path.join("dir1", "file2"),
+            os.path.join("dir1", "file3"),
+            os.path.join("dir1", "dvc_file2"),
+            os.path.join("dir1", "dvc_file2.dvc"),
+        ]
+    )
+
+
+def test_ls_repo_with_wrong_target(tmp_dir, dvc, scm):
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+
+    with pytest.raises(DvcException):
+        Repo.ls(fspath(tmp_dir), target="wrong_target")
+
+
+def test_ls_repo_with_wrong_target_outputs_only(tmp_dir, dvc, scm):
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+
+    with pytest.raises(DvcException):
+        Repo.ls(
+            fspath(tmp_dir),
+            target="wrong_target",
+            recursive=True,
+            outputs_only=True,
+        )
+
+
+def test_ls_repo_with_file_target(tmp_dir, dvc, scm):
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+
+    files = Repo.ls(
+        fspath(tmp_dir), target=os.path.join("dir1", "file2"), recursive=True
+    )
+    assert set(files) == set([os.path.join("dir1", "file2")])
+
+
+def test_ls_repo_with_removed_dvc_dir(tmp_dir, dvc, scm):
+    script = os.linesep.join(
+        [
+            "from pathlib import Path",
+            "Path({}).touch()".format(os.path.join("out", "file")),
+        ]
+    )
+    tmp_dir.scm_gen({"script.py": script}, commit="init")
+    tmp_dir.dvc_gen({"dep": "content"}, commit="init dvc")
+    dvc.run(
+        **{
+            "command": "python script.py",
+            "outs": [os.path.join("out", "file")],
+            "deps": ["dep"],
+            "fname": "out.dvc",
+        }
+    )
+    tmp_dir.scm_add(["out.dvc"], commit="run")
+    shutil.rmtree("out")
+
+    files = Repo.ls(fspath(tmp_dir), target="out", recursive=True)
+    assert set(files) == set([os.path.join("out", "file")])
+
+
+def test_ls_repo_remote(tmp_dir, dvc, scm, mocker):
+    from dvc.scm import Git
+    from distutils.dir_util import copy_tree
+
+    tmp_dir.scm_gen(FS_STRUCTURE, commit="init")
+    tmp_dir.dvc_gen(DVC_STRUCTURE, commit="dvc")
+    m = mocker.patch("dvc.scm.Git.clone")
+
+    def clone_repo(url, to_path, *args):
+        assert url == "git://github.com:repo"
+        copy_tree(fspath(tmp_dir), to_path)
+        return Git(fspath(tmp_dir))
+
+    m.side_effect = clone_repo
+
+    files = Repo.ls("git://github.com:repo")
```

Comment:
You can use `erepo_dir` fixture to create an external dvcrepo and `git_dir` for external git-only repository. If you want to test with url, you can use `file://` based urls as well. No need to mock.

Reply:
Not sure if `file://` works these days. In any case, the logic still passes this to `external_repo`, so we could use local paths here with the same effect. I know that previously we've had a different implementation here for which the url in tests mattered, but it is no longer like that, so we could simplify and use `erepo_dir` as @skshetry suggested :slightly_smiling_face:

Reply:
@Suor always adds `file://` for external repo in combination with `erepo_dir` with`# Use URL to prevent any fishy optimizations` kind-of comments. 

https://github.com/iterative/dvc/blob/7821fa6866a492c0675248028039257d292a09ec/tests/func/test_external_repo.py#L48-L50